### ISSUE TITLE
[BUG] - Root servers update failed

### ISSUE THREAD

**Issue by Cebeerre** [semihalev/sdns#253]
### Description

I'm happily running sdns on top of OpenWRT as the DNS upstream for AdGuardHome. From time to time, I'm seeing the following message in the log:

`Wed Dec  6 03:46:26 2023 daemon.info sdns[17660]: t=2023-12-06T02:46:26+0000 lvl=eror msg="Root servers update failed" error="A record count mismatch"`

I've been taking a look at the code, but I'm not really familiar with Go, so I'm not sure what the purpose of the snippet where appears this comparison: `ipv4Count != nsCount` 

Is this something really expected ?

Thanks !

### Reproduction steps

```bash
I guess this `checkPriming` thing is scheduled, so it happens from time to time.
```

### sdns version

```bash
1.3.5
```

### Logs

```bash
Wed Dec  6 03:46:26 2023 daemon.info sdns[17660]: t=2023-12-06T02:46:26+0000 lvl=eror msg="Root servers update failed" error="A record count mismatch"
```

### OS

Linux

### Additional context

I'm using the default configuration for the root servers. Please note that my ISP hasn't enabled IPv6 yet, so currently using just IPv4 (I left the IPv6 addresses for the root servers though). sdns is listening just on localhost

```
# Address to bind to for the DNS server.
bind = "127.0.0.1:5354"
```

**Comment by semihalev** [comment]
Thanks for the report and details.

Are you using any forwarder?

**Comment by Cebeerre** [comment]
Not really, forwarders and fallback are empty:

```
# Failover resolver IPv4 or IPv6 addresses with port, left blank for disabled.
# fallbackservers = [
#   "8.8.8.8:53",
#   "[2001:4860:4860::8888]:53"
# ]
fallbackservers = [
]

# Forwarder resolver IPv4 or IPv6 addresses with port, left blank for disabled.
# forwarderservers = [
#   "8.8.8.8:53",
#   "[2001:4860:4860::8888]:53",
#   "tls://8.8.8.8:853"
# ]
forwarderservers = [
]
```
Thanks !!

**Comment by semihalev** [comment]
Normally, known root servers count is `13`.  SDNS checks glue records from a root server for IPv4 and IPv6 address count with the priming queries. You can see the `dig` result in the below, on the additional section there should be `13` IPv4 address. This is small compare from the response. I think on your side one or more root dns servers are giving less glue records on additional section. This shouldn't happen. On my personel server, I never get that error more than 2 years. But this error never corrupt current root servers on your server. This mean, SDNS will be keep old corrected root servers. 

Simply, can you please test the glue records from root servers with `dig` command? We may find the problem with this test.

```
# dig ns . @a.root-servers.net

; <<>> DiG 9.19.12 <<>> ns . @a.root-servers.net
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 2126
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;.				IN	NS

;; ANSWER SECTION:
.			518400	IN	NS	e.root-servers.net.
.			518400	IN	NS	h.root-servers.net.
.			518400	IN	NS	l.root-servers.net.
.			518400	IN	NS	i.root-servers.net.
.			518400	IN	NS	a.root-servers.net.
.			518400	IN	NS	d.root-servers.net.
.			518400	IN	NS	c.root-servers.net.
.			518400	IN	NS	b.root-servers.net.
.			518400	IN	NS	j.root-servers.net.
.			518400	IN	NS	k.root-servers.net.
.			518400	IN	NS	g.root-servers.net.
.			518400	IN	NS	m.root-servers.net.
.			518400	IN	NS	f.root-servers.net.

;; ADDITIONAL SECTION:
e.root-servers.net.	518400	IN	A	192.203.230.10
e.root-servers.net.	518400	IN	AAAA	2001:500:a8::e
h.root-servers.net.	518400	IN	A	198.97.190.53
h.root-servers.net.	518400	IN	AAAA	2001:500:1::53
l.root-servers.net.	518400	IN	A	199.7.83.42
l.root-servers.net.	518400	IN	AAAA	2001:500:9f::42
i.root-servers.net.	518400	IN	A	192.36.148.17
i.root-servers.net.	518400	IN	AAAA	2001:7fe::53
a.root-servers.net.	518400	IN	A	198.41.0.4
a.root-servers.net.	518400	IN	AAAA	2001:503:ba3e::2:30
d.root-servers.net.	518400	IN	A	199.7.91.13
d.root-servers.net.	518400	IN	AAAA	2001:500:2d::d
c.root-servers.net.	518400	IN	A	192.33.4.12
c.root-servers.net.	518400	IN	AAAA	2001:500:2::c
b.root-servers.net.	518400	IN	A	170.247.170.2
b.root-servers.net.	518400	IN	AAAA	2801:1b8:10::b
j.root-servers.net.	518400	IN	A	192.58.128.30
j.root-servers.net.	518400	IN	AAAA	2001:503:c27::2:30
k.root-servers.net.	518400	IN	A	193.0.14.129
k.root-servers.net.	518400	IN	AAAA	2001:7fd::1
g.root-servers.net.	518400	IN	A	192.112.36.4
g.root-servers.net.	518400	IN	AAAA	2001:500:12::d0d
m.root-servers.net.	518400	IN	A	202.12.27.33
m.root-servers.net.	518400	IN	AAAA	2001:dc3::35
f.root-servers.net.	518400	IN	A	192.5.5.241
f.root-servers.net.	518400	IN	AAAA	2001:500:2f::f

;; Query time: 56 msec
;; SERVER: 2001:503:ba3e::2:30#53(a.root-servers.net) (UDP)
;; WHEN: Wed Dec 06 10:53:18 +03 2023
;; MSG SIZE  rcvd: 811

**Comment by Cebeerre** [comment]
Sure thing, there you go:

```
root@router:/etc# dig ns . @a.root-servers.net

; <<>> DiG 9.18.19 <<>> ns . @a.root-servers.net
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 15587
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;.				IN	NS

;; ANSWER SECTION:
.			518400	IN	NS	e.root-servers.net.
.			518400	IN	NS	h.root-servers.net.
.			518400	IN	NS	l.root-servers.net.
.			518400	IN	NS	i.root-servers.net.
.			518400	IN	NS	a.root-servers.net.
.			518400	IN	NS	d.root-servers.net.
.			518400	IN	NS	c.root-servers.net.
.			518400	IN	NS	b.root-servers.net.
.			518400	IN	NS	j.root-servers.net.
.			518400	IN	NS	k.root-servers.net.
.			518400	IN	NS	g.root-servers.net.
.			518400	IN	NS	m.root-servers.net.
.			518400	IN	NS	f.root-servers.net.

;; ADDITIONAL SECTION:
e.root-servers.net.	518400	IN	A	192.203.230.10
e.root-servers.net.	518400	IN	AAAA	2001:500:a8::e
h.root-servers.net.	518400	IN	A	198.97.190.53
h.root-servers.net.	518400	IN	AAAA	2001:500:1::53
l.root-servers.net.	518400	IN	A	199.7.83.42
l.root-servers.net.	518400	IN	AAAA	2001:500:9f::42
i.root-servers.net.	518400	IN	A	192.36.148.17
i.root-servers.net.	518400	IN	AAAA	2001:7fe::53
a.root-servers.net.	518400	IN	A	198.41.0.4
a.root-servers.net.	518400	IN	AAAA	2001:503:ba3e::2:30
d.root-servers.net.	518400	IN	A	199.7.91.13
d.root-servers.net.	518400	IN	AAAA	2001:500:2d::d
c.root-servers.net.	518400	IN	A	192.33.4.12
c.root-servers.net.	518400	IN	AAAA	2001:500:2::c
b.root-servers.net.	518400	IN	A	170.247.170.2
b.root-servers.net.	518400	IN	AAAA	2801:1b8:10::b
j.root-servers.net.	518400	IN	A	192.58.128.30
j.root-servers.net.	518400	IN	AAAA	2001:503:c27::2:30
k.root-servers.net.	518400	IN	A	193.0.14.129
k.root-servers.net.	518400	IN	AAAA	2001:7fd::1
g.root-servers.net.	518400	IN	A	192.112.36.4
g.root-servers.net.	518400	IN	AAAA	2001:500:12::d0d
m.root-servers.net.	518400	IN	A	202.12.27.33
m.root-servers.net.	518400	IN	AAAA	2001:dc3::35
f.root-servers.net.	518400	IN	A	192.5.5.241
f.root-servers.net.	518400	IN	AAAA	2001:500:2f::f

;; Query time: 39 msec
;; SERVER: 198.41.0.4#53(a.root-servers.net) (UDP)
;; WHEN: Wed Dec 06 09:13:05 CET 2023
;; MSG SIZE  rcvd: 811
```
But I'm actually getting the same result as you ...

**Comment by semihalev** [comment]
From others b, c, d, e, f, g, h, i, j ,k, l and m?

**Comment by Cebeerre** [comment]
Yes, sorry, I tested as well all the others ... same thing.

**Comment by semihalev** [comment]
Thanks the tests. This is interesting. I will try investigate the problem.

**Comment by Cebeerre** [comment]
All of them report 13 ANSWERS as expected:

```
root@router:/etc# root_servers=$(curl -s https://www.iana.org/domains/root/servers)
root@router:/etc# server_names=$(echo "$root_servers" | grep -Eo '[a-z]\.root-servers\.net'| sed 's/\.$//')
root@router:/etc# for server in $server_names
> do
>     dig ns . @$server | grep "ANSWER:"
> done
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
;; flags: qr aa rd; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27
root@router:/etc#
```
Thanks !!

**Comment by Cebeerre** [comment]
Interestingly, doing the same thing but going through the route AdGuardHome -> sdns, it actually reports 13 NS servers, but the additional section reports 24 IPs.

```
root@router:/etc# dig ns .

; <<>> DiG 9.18.19 <<>> ns .
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 1046
;; flags: qr rd ra; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 24

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;.				IN	NS

;; ANSWER SECTION:
.			346046	IN	NS	a.root-servers.net.
.			346046	IN	NS	b.root-servers.net.
.			346046	IN	NS	c.root-servers.net.
.			346046	IN	NS	d.root-servers.net.
.			346046	IN	NS	e.root-servers.net.
.			346046	IN	NS	f.root-servers.net.
.			346046	IN	NS	g.root-servers.net.
.			346046	IN	NS	h.root-servers.net.
.			346046	IN	NS	i.root-servers.net.
.			346046	IN	NS	j.root-servers.net.
.			346046	IN	NS	k.root-servers.net.
.			346046	IN	NS	l.root-servers.net.
.			346046	IN	NS	m.root-servers.net.

;; ADDITIONAL SECTION:
a.root-servers.net.	346046	IN	A	198.41.0.4
a.root-servers.net.	346046	IN	AAAA	2001:503:ba3e::2:30
b.root-servers.net.	346046	IN	A	170.247.170.2
b.root-servers.net.	346046	IN	AAAA	2801:1b8:10::b
c.root-servers.net.	346046	IN	A	192.33.4.12
c.root-servers.net.	346046	IN	AAAA	2001:500:2::c
d.root-servers.net.	346046	IN	A	199.7.91.13
d.root-servers.net.	346046	IN	AAAA	2001:500:2d::d
e.root-servers.net.	346046	IN	A	192.203.230.10
e.root-servers.net.	346046	IN	AAAA	2001:500:a8::e
f.root-servers.net.	346046	IN	A	192.5.5.241
f.root-servers.net.	346046	IN	AAAA	2001:500:2f::f
g.root-servers.net.	346046	IN	A	192.112.36.4
g.root-servers.net.	346046	IN	AAAA	2001:500:12::d0d
h.root-servers.net.	346046	IN	A	198.97.190.53
h.root-servers.net.	346046	IN	AAAA	2001:500:1::53
i.root-servers.net.	346046	IN	A	192.36.148.17
i.root-servers.net.	346046	IN	AAAA	2001:7fe::53
j.root-servers.net.	346046	IN	A	192.58.128.30
j.root-servers.net.	346046	IN	AAAA	2001:503:c27::2:30
k.root-servers.net.	346046	IN	A	193.0.14.129
k.root-servers.net.	346046	IN	AAAA	2001:7fd::1
l.root-servers.net.	346046	IN	A	199.7.83.42

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1) (UDP)
;; WHEN: Wed Dec 06 09:46:04 CET 2023
;; MSG SIZE  rcvd: 739

```

Querying sdns directly, reports 27 IPs:

```
root@router:/etc# dig ns . @127.0.0.1 -p 5354

; <<>> DiG 9.18.19 <<>> ns . @127.0.0.1 -p 5354
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 2057
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 27

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
; COOKIE: e6b1a323c2c1dc4a012adb8c0f2d41c68d09a41e5cbf44d31a5e759f1993dff73de54b9a9fbd6725 (good)
;; QUESTION SECTION:
;.				IN	NS

;; ANSWER SECTION:
.			86377	IN	NS	g.root-servers.net.
.			86377	IN	NS	e.root-servers.net.
.			86377	IN	NS	j.root-servers.net.
.			86377	IN	NS	k.root-servers.net.
.			86377	IN	NS	b.root-servers.net.
.			86377	IN	NS	m.root-servers.net.
.			86377	IN	NS	a.root-servers.net.
.			86377	IN	NS	d.root-servers.net.
.			86377	IN	NS	h.root-servers.net.
.			86377	IN	NS	f.root-servers.net.
.			86377	IN	NS	i.root-servers.net.
.			86377	IN	NS	c.root-servers.net.
.			86377	IN	NS	l.root-servers.net.

;; ADDITIONAL SECTION:
m.root-servers.net.	86377	IN	A	202.12.27.33
l.root-servers.net.	86377	IN	A	199.7.83.42
k.root-servers.net.	86377	IN	A	193.0.14.129
j.root-servers.net.	86377	IN	A	192.58.128.30
i.root-servers.net.	86377	IN	A	192.36.148.17
h.root-servers.net.	86377	IN	A	198.97.190.53
g.root-servers.net.	86377	IN	A	192.112.36.4
f.root-servers.net.	86377	IN	A	192.5.5.241
e.root-servers.net.	86377	IN	A	192.203.230.10
d.root-servers.net.	86377	IN	A	199.7.91.13
c.root-servers.net.	86377	IN	A	192.33.4.12
b.root-servers.net.	86377	IN	A	170.247.170.2
a.root-servers.net.	86377	IN	A	198.41.0.4
m.root-servers.net.	86377	IN	AAAA	2001:dc3::35
l.root-servers.net.	86377	IN	AAAA	2001:500:9f::42
k.root-servers.net.	86377	IN	AAAA	2001:7fd::1
j.root-servers.net.	86377	IN	AAAA	2001:503:c27::2:30
i.root-servers.net.	86377	IN	AAAA	2001:7fe::53
h.root-servers.net.	86377	IN	AAAA	2001:500:1::53
g.root-servers.net.	86377	IN	AAAA	2001:500:12::d0d
f.root-servers.net.	86377	IN	AAAA	2001:500:2f::f
e.root-servers.net.	86377	IN	AAAA	2001:500:a8::e
d.root-servers.net.	86377	IN	AAAA	2001:500:2d::d
c.root-servers.net.	86377	IN	AAAA	2001:500:2::c
b.root-servers.net.	86377	IN	AAAA	2801:1b8:10::b
a.root-servers.net.	86377	IN	AAAA	2001:503:ba3e::2:30

;; Query time: 0 msec
;; SERVER: 127.0.0.1#5354(127.0.0.1) (UDP)
;; WHEN: Wed Dec 06 09:48:56 CET 2023
;; MSG SIZE  rcvd: 855
```
But I assume sdns is querying directly the root servers, so it's not using the DNS configured in the resolv.conf, right ?

Thanks !

**Comment by semihalev** [comment]
Yes, sdns is querying directly from root servers, If any forwarder is not defined.

**Comment by semihalev** [comment]
I released v1.3.6 and dnssec onoff capability added there. I'm going to close the issue and if you have any other related issues, you can open another one.
